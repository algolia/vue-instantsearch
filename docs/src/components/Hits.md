---
title: Hits
mainTitle: Components
layout: main.pug
category: Components
withHeadings: true
navWeight: 6
editable: true
githubSource: docs/src/components/hits.md
---

A component to format and render the search results.

<a class="btn btn-static-theme" href="stories/?selectedKind=Hits">ðŸ•¹ try out live</a>

## Usage

When you want to override how a single item looks:

```html
<ais-hits>
  <template slot="item" slot-scope="{ item }">
    <h2>
      <a :href="item.url">
        {{ item.title }}
      </a>
    </h2>
    <p>{{ item.description }}</p>
  </template>
</ais-hits>
```

When you want to override the complete hits:

```html
<ais-hits>
  <template slot="default" slot-scope="{ items }">
    <div v-for="(item, itemIndex) in items" :key="itemIndex">
      <h2>
        <a :href="item.url">
          {{ item.title }}
        </a>
      </h2>
      <p>{{ item.description }}</p>
    </div>
  </template>
</ais-hits>
```

When you want to show something custom if there are no results:

```html
<ais-hits>
  <template slot="default" slot-scope="{ items }">
    <div v-if="items.length <= 0">
      No results found.
    </div>
    <div v-else v-for="(item, itemIndex) in items" :key="itemIndex">
      <h2>
        <a :href="item.url">
          {{ item.title }}
        </a>
      </h2>
      <p>{{ item.description }}</p>
    </div>
  </template>
</ais-hits>
```

## Props

Name | Type | Default | Description | Required
---|---|---|---|---
escapeHTML | Boolean | true | Escape raw HTML in the hits | no
transformItems | `(items: object[]) => object[]` | `x => x` | Function which receives the items, which will be called before displaying them. Should return a new array with the same shape as the original array. Useful for mapping over the items to transform, remove or reorder them | -
classNames | Object | | Override class names | no

## CSS classes

Here's a list of CSS classes exposed by this widget. To better understand the underlying
DOM structure, have a look at the generated DOM in your browser.

Note that you can pass the prop `class-names`, with an object of class names and their replacement to override this.

Class name | Description
---|---
`ais-Hits` | Container class
`ais-Hits-list` | An item
`ais-Hits-item` | An item
