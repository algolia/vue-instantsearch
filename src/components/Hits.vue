<template>
  <div :class="suit()" v-if="state">
    <slot :items="items">
      <ol :class="suit('list')">
        <li
          v-for="(item, itemIndex) in items"
          :key="item.objectID"
          :class="suit('item')"
        >
          <slot name="item" :item="item" :index="itemIndex">
            objectID: {{item.objectID}}, index: {{itemIndex}}
          </slot>
        </li>
      </ol>
    </slot>
  </div>
</template>

<script>
import algoliaComponent from '../component';
import { connectHits } from 'instantsearch.js/es/connectors';

export default {
  mixins: [algoliaComponent],
  data() {
    return {
      widgetName: 'Hits',
    };
  },
  beforeCreate() {
    this.connector = connectHits;
  },
  computed: {
    // Fixes InstantSearch.js connectors API: every list of things must be called `items`
    items() {
      return this.state.hits;
    },
  },
};</script>
