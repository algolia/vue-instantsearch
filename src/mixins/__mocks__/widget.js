let state = {};

// we need to have state given by `component` before it is mounted, otherwise
// we can't render it in most cases (items, hits, etc. are used in the template)
// so we share a (mock) global state during a whole test.
//
// (a mock is imported once per test file, so the state is isolated between tests)
//
// This allows us to import this `__setState` function and call it in the test
// to give the necessary data before mounting.
export function __setState(newState) {
  state = newState;
}

export const createWidgetMixin = () => ({
  data() {
    return {
      state,
    };
  },
});
