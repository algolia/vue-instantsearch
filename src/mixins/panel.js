import Vue from 'vue';

export const PANEL_EMITTER_NAMESPACE = 'instantSearchPanelEmitter';
export const PANEL_CHANGE_EVENT = 'PANEL_CHANGE_EVENT';

export const createPanelProviderMixin = () => ({
  props: {
    emitter: {
      type: Object,
      required: false,
      default() {
        return new Vue({
          name: 'PanelProvider',
        });
      },
    },
  },
  provide() {
    return {
      [PANEL_EMITTER_NAMESPACE]: this.emitter,
    };
  },
  data() {
    return {
      canRefine: true,
    };
  },
  created() {
    this.emitter.$on(PANEL_CHANGE_EVENT, value => {
      this.updateCanRefine(value);
    });
  },
  beforeDestroy() {
    this.emitter.$destroy();
  },
  methods: {
    updateCanRefine(value) {
      this.canRefine = value;
    },
  },
});

export const createPanelConsumerMixin = ({ mapStateToCanRefine }) => ({
  inject: {
    emitter: {
      from: PANEL_EMITTER_NAMESPACE,
      default() {
        return {
          $emit: () => {},
        };
      },
    },
  },
  data() {
    return {
      state: null,
      hasAlreadyEmitted: false,
    };
  },
  watch: {
    state: {
      immediate: true,
      handler(nextState, previousState) {
        if (!nextState) {
          return;
        }

        const previousCanRefine = mapStateToCanRefine(previousState || {});
        const nextCanRefine = mapStateToCanRefine(nextState);

        if (!this.hasAlreadyEmitted || previousCanRefine !== nextCanRefine) {
          this.emitter.$emit(PANEL_CHANGE_EVENT, nextCanRefine);

          this.hasAlreadyEmitted = true;
        }
      },
    },
  },
});
