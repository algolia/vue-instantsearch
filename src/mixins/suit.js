import suit from '../util/suit';

export const createSuitMixin = ({ name }) => ({
  props: {
    classNames: {
      type: Object,
      default: undefined,
    },
  },
  methods: {
    suit(element, modifier) {
      const className = suit(name, element, modifier);
      const userClassName = this.classNames && this.classNames[className];
      if (userClassName) {
        return [className, userClassName].join(' ');
      }
      return className;
    },
  },
});
