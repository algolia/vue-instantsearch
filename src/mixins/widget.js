const defaultShouldUpdate = (previous, next) => previous !== next;

export const createWidgetMixin = ({ connect }) => ({
  inject: {
    instantSearchInstance: {
      name: 'instantSearchInstance',
      default() {
        const tag = this.$options._componentTag;
        throw new TypeError(
          `It looks like you forgot to wrap your Algolia search component
          "<${tag}>" inside of an "<ais-index>" component.`
        );
      },
    },
  },
  data() {
    return {
      state: null,
    };
  },
  computed: {
    widgetParams() {
      const { widgetParams } = this.$options;

      return Object.keys(widgetParams).reduce(
        (acc, name) => ({
          ...acc,
          [name]: widgetParams[name].getValue(this),
        }),
        {}
      );
    },
  },
  watch: {
    widgetParams(next, previous) {
      const { widgetParams } = this.$options;

      const shouldUpdateWidget = Object.keys(widgetParams).some(name => {
        const { shouldUpdate = defaultShouldUpdate } = widgetParams[name];

        return shouldUpdate(previous[name], next[name]);
      });

      if (shouldUpdateWidget) {
        this.instantSearchInstance.removeWidget(this.widget);
        this.widget = this.widgetFactory(next);
        this.instantSearchInstance.addWidget(this.widget);
      }
    },
  },
  created() {
    this.widgetFactory = connect(this.updateData, () => {});
    this.widget = this.widgetFactory(this.widgetParams);
    this.instantSearchInstance.addWidget(this.widget);
  },
  beforeDestroy() {
    this.instantSearchInstance.removeWidget(this.widget);
  },
  methods: {
    updateData(state = {}) {
      this.state = state;
    },
  },
});
