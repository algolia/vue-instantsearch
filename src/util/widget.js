export const addWidget = (connector, parent, widgetParams, render) => {
  const factory = connector(render);
  const widget = factory(widgetParams);
  parent.addWidgets([widget]);

  return widget;
};

export const forceRender = (widget, parent, instantSearchInstance) => {
  if (
    instantSearchInstance.__initialSearchResults &&
    !instantSearchInstance.started
  ) {
    if (typeof instantSearchInstance.__forceRender !== 'function') {
      throw new Error(
        'You are using server side rendering with <ais-instant-search> instead of <ais-instant-search-ssr>.'
      );
    }
    instantSearchInstance.__forceRender(widget, parent);
  }
};

export const removeWidget = (widget, parent) => {
  if (widget) {
    parent.removeWidgets([widget]);
  }
};
