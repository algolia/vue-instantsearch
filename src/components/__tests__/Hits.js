import { mount } from '@vue/test-utils';
import { __setState } from '../../mixins/widget';
import Hits from '../Hits.vue';

jest.mock('../../mixins/widget');

const defaultState = {
  hits: [{ objectID: 'one' }, { objectID: 'two' }],
};

it('accepts an escapeHTML prop', () => {
  __setState({
    ...defaultState,
  });

  const wrapper = mount(Hits, {
    propsData: {
      escapeHTML: true,
    },
  });

  expect(wrapper.vm.widgetParams.escapeHTML).toBe(true);
});

it('accepts a transformItems prop', () => {
  __setState({
    ...defaultState,
  });

  const transformItems = () => {};

  const wrapper = mount(Hits, {
    propsData: {
      transformItems,
    },
  });

  expect(wrapper.vm.widgetParams.transformItems).toBe(transformItems);
});

it('renders correctly', () => {
  __setState({
    ...defaultState,
  });

  const wrapper = mount(Hits);

  expect(wrapper.html()).toMatchSnapshot();
});

it('exposes insights prop to the default slot', () => {
  const insights = jest.fn();
  __setState({
    ...defaultState,
    insights,
  });

  const wrapper = mount(Hits, {
    scopedSlots: {
      default: `
        <ul slot-scope="{ items, insights }">
          <li v-for="(item, itemIndex) in items" >
          
            <button :id="'add-to-cart-' + item.objectID" @click="insights('clickedObjectIDsAfterSearch', {eventName: 'Add to cart', objectIDs: [item.objectID]})">
              Add to cart
            </button>
          </li>
        </ul>
      `,
    },
  });
  wrapper.find('#add-to-cart-two').trigger('click');
  expect(insights).toHaveBeenCalledWith('clickedObjectIDsAfterSearch', {
    eventName: 'Add to cart',
    objectIDs: ['two'],
  });
});

it('exposes insights prop to the item slot', () => {
  const insights = jest.fn();
  __setState({
    ...defaultState,
    insights,
  });

  const wrapper = mount(Hits, {
    scopedSlots: {
      item: `
        <div slot-scope="{ item, insights }">
          <button :id="'add-to-cart-' + item.objectID" @click="insights('clickedObjectIDsAfterSearch', {eventName: 'Add to cart', objectIDs: [item.objectID]})">
            Add to cart
          </button>
        </div>
      `,
    },
  });
  wrapper.find('#add-to-cart-two').trigger('click');
  expect(insights).toHaveBeenCalledWith('clickedObjectIDsAfterSearch', {
    eventName: 'Add to cart',
    objectIDs: ['two'],
  });
});
