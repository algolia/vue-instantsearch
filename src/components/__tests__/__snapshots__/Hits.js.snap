// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `

<div class="ais-Hits">
  <ol class="ais-Hits-list">
    <li class="ais-Hits-item">
      objectID: one, index: 0
    </li>
    <li class="ais-Hits-item">
      objectID: two, index: 1
    </li>
  </ol>
</div>

`;
